app/util/ReferrerUrlChecking.scala

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package util
+
+import config.{AppConfig, CFConfig}
+import play.api.Logging
+import play.api.mvc.{AnyContent, Request}
+import uk.gov.hmrc.play.bootstrap.binders.RedirectUrl._
+import uk.gov.hmrc.play.bootstrap.binders._
+
+trait ReferrerUrlChecking extends Logging {
+
+  implicit class OptionStringOps(opt: Option[String]) {
+    private val notApplicable               = "n/a"
+    def orNotApplicable: Option[String]     = Some(opt.getOrElse(notApplicable))
+    def removeNotApplicable: Option[String] = opt match {
+      case Some(referrer) if referrer == notApplicable => None
+      case _                                           => opt
+    }
+  }
+
+  def logInvalidUrl(service: Option[String], url: RedirectUrl)(implicit appConfig: AppConfig): SafeRedirectUrl =
+    url.getEither(appConfig.urlPolicy) match {
+      case Right(safeRedirectUrl: SafeRedirectUrl) => safeRedirectUrl
+      case Left(unsafeRedirectError)               =>
+        val serviceName = if (service.isEmpty) "A service" else s"Service [${service.get}]"
+        logger.warn(s"$serviceName is attempting to redirect to an unsafe url: $unsafeRedirectError")
+        RedirectUrl(url.unsafeValue).get(UnsafePermitAll)
+    }
```

Comment:
feels like this trait is doing more than it should, and I'm not sure about the naming - logInvalidUrl is something it might do when it's doing it's job of getting a SafeRedirectUrl

the other logic in this file "maybeReferrerAsString" feels like it's possible got some context specific stuff in it - I'd opt for doing this in smaller steps - so leave this stuff where it is at the moment, and only implement the mapping for now of RedirectUrl => SafeRedirectUrl => String then maybe have a tech debt item to circle back (or even just in another pr on same ticket) extract another helper for other repeated functionality